### PR TITLE
Publish statistics through MetricQ

### LINKED ISSUE #19: Send db statistics as metricq source
We dump some statistics in the text log... I wonder if there is a better way?

### PR DESCRIPTION
**Pull request by tilsche**
Briefly tested it and it seemed to work so far.

Unfortunately the wizard can't configure the created metrics because their source is a db... So that will be another issue.

fixes #19

### CODE REVIEW

File: src/db_stats.cpp [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        assert(duration > 0);
+        request_rate_.send({ time, stats.ready_count_ / duration });
+        data_rate_.send({ time, stats.data_size_ / duration });
+        pending_time_.send({ time, std::chrono::duration_cast<std::chrono::duration<double>>(
+                                       stats.pending_duration_)
+                                       .count() });
+        active_time_.send({ time, std::chrono::duration_cast<std::chrono::duration<double>>(
+                                      stats.active_duration_)
+                                      .count() });
+        pending_count_.send({ time, (double)stats.pending_count_ });
```
Comment (bmario) [inline src/db_stats.cpp]: Don't C-style cast

File: src/db_stats.cpp [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        pending_count_.metadata.unit("");
+        pending_count_.metadata.quantity("");
+        pending_count_.metadata.description(
+            fmt::format("number of pending {}-requests", read_or_write));
+        pending_count_.metadata.scope(metricq::Metadata::Scope::point);
+        pending_count_.metadata.rate(rate);
+
+        active_count_.metadata.unit("");
+        active_count_.metadata.quantity("");
+        active_count_.metadata.description("number of actively processed {}-requests");
```
Comment (bmario) [inline src/db_stats.cpp]: Looks like a missing fmt::format or something.

File: src/db_stats.cpp [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    {
+    }
+
+    void collect()
+    {
+        auto time = metricq::Clock::now();
+        double duration =
+            std::chrono::duration_cast<std::chrono::duration<double>>(time - previous_collect_time_)
+                .count();
+        auto read_stats = read.collect();
```
Comment (bmario) [inline src/db_stats.cpp]: I feel like these locals should be inline instead.

File: src/db_stats.cpp [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+        std::lock_guard lock(stats_mutex_);
+        Stats collected_stats = stats_;
+        stats_ = Stats();
+        return collected_stats;
+    }
+
+private:
+    std::mutex stats_mutex_;
+    Stats stats_;
+    metricq::TimePoint previous_dump_time_;
```
Comment (bmario) [inline src/db_stats.cpp]: This variable is never read and only set once in the constructor.

**Review by bmario — CHANGES_REQUESTED** [review]

**Review by bmario — COMMENTED** [review]
Also, I haven't seen a `declare_metrics()` call.

**Comment by tilsche** [comment]
> Also, I haven't seen a `declare_metrics()` call.

That's done in the base class
https://github.com/metricq/metricq-cpp/blob/experimental-async-cpp-db/src/db.cpp#L199

**Comment by tilsche** [comment]
I believe I have addressed all the review comments.

**Review by bmario — APPROVED** [review]

**Pull request merged**